src/probability/independence.lean

```diff
@@ -309,6 +308,68 @@ begin
   exact indep_sets.indep_aux h2 hp2 hpm2 hyp ht ht2,
 end
 
+variables {α ι : Type*} {m0 : measurable_space α} {μ : measure α}
+
+lemma pi_system_indep_insert {π : ι → set (set α)} {a : ι} {S : finset ι}
+  (hp_ind : Indep_sets π μ) (haS : a ∉ S) :
+  indep_sets (pi_Union_Inter π {S}) (π a) μ :=
+begin
+  rintros t1 t2 ⟨s, hs_mem, ft1, hft1_mem, ht1_eq⟩ ht2_mem_pia,
+  rw set.mem_singleton_iff at hs_mem,
+  subst hs_mem,
+  let f := λ n, ite (n = a) t2 (ite (n ∈ s) (ft1 n) set.univ),
+  have h_f_mem : ∀ n ∈ insert a s, f n ∈ π n,
+  { intros n hn_mem_insert,
+    simp_rw f,
+    cases (finset.mem_insert.mp hn_mem_insert) with hn_mem hn_mem,
+    { simp [hn_mem, ht2_mem_pia], },
+    { have hn_ne_a : n ≠ a, by { rintro rfl, exact haS hn_mem, },
+      simp [hn_ne_a, hn_mem, hft1_mem n hn_mem], }, },
+  have h_f_mem_pi : ∀ n ∈ s, f n ∈ π n, from λ x hxS, h_f_mem x (by simp [hxS]),
+  have h_t1 : t1 = ⋂ n ∈ s, f n,
+  { suffices h_forall : ∀ n ∈ s, f n = ft1 n,
+    { rw ht1_eq,
+      congr' with n x,
+      congr' with hns y,
+      simp only [(h_forall n hns).symm], },
+    intros n hnS,
+    have hn_ne_a : n ≠ a, by { rintro rfl, exact haS hnS, },
+    simp_rw [f, if_pos hnS, if_neg hn_ne_a], },
+  have h_μ_t1 : μ t1 = ∏ n in s, μ (f n), by rw [h_t1, ←hp_ind s h_f_mem_pi],
+  have h_t2 : t2 = f a, by { simp_rw [f], simp, },
+  have h_μ_inter : μ (t1 ∩ t2) = ∏ n in insert a s, μ (f n),
+  { have h_t1_inter_t2 : t1 ∩ t2 = ⋂ n ∈ insert a s, f n,
+      by rw [h_t1, h_t2, finset.set_bInter_insert, set.inter_comm],
+    rw [h_t1_inter_t2, ←hp_ind (insert a s) h_f_mem], },
+  rw [h_μ_inter, finset.prod_insert haS, h_t2, mul_comm, h_μ_t1],
+end
+
+/-- The measurable space structures generated by independent pi-systems are independent. -/
+theorem Indep_sets.Indep [is_probability_measure μ] (m : ι → measurable_space α)
+  (h_le : ∀ i, m i ≤ m0) (π : ι → set (set α)) (h_pi : ∀ n, is_pi_system (π n))
+  (hp_univ : ∀ i, set.univ ∈ π i) (h_generate : ∀ i, m i = generate_from (π i))
```

Comment:
It is not obvious to me that `hp_univ` is really needed here. Can you comment on that?